Fixed a bug in TFFunctionPartition::markBlock() when handling a loop.

mhong opened the pull request:
The is an "off-by-1" error -- we need to check if `BB` **properly** post-dominates `pred`, instead of mere post-dominance, since `BB` and `pred` can be the same block, as in the test case, where the CFG is: 
```
bb0 -> bb1 -> {bb2, bb3}; bb2 -> bb1
```
Here bb1 is the loop header, bb2 the loop body.

In this case, when we run `markBlock()` over bb1, we first find bb2,
post-dominated by bb1. We then walk bb2 (pushing it to `worklist`), where we find
bb1 as its predecessor -- in that case we must mark the term inst of bb1 (a
cond_br), because all tensor ops in bb1 are control-dependent on that term inst. This way, the loop count variable gets promoted to tensor.

The comment blocks in the code and added test cases provide more context.
Resolves #50723.

Additional misc changes:
1. Renamed `visited` to `addedToList` to better reflect its semantics, and
moved around sanity check code on `numTensorSuccs` to make it clear that the
core marking logic does not depend on it.

2. Added more sanity-checking to compiler runtime, to catch incorrect code generation by the compiler like the case at hand. e.g. if `@_swift_tfc_StartTensorComputation()` is called more than once on a tensor computation handle, the runtime will crash with an error message like:
```
Fatal error: finish() is called in async execution mode with pthread == nil -- Was finish() already called?: file /usr/local/google/home/hongm/ssd_part/git/swift-base/swift/stdlib/public/TensorFlow/CompilerRuntime.swift, line 768
```

Commit:
Fixed a bug in TFFunctionPartition::markBlock() when handling a loop.

An example CFG: bb0 -> bb1 -> {bb2, bb3}; bb2 -> bb1
Here bb1 is the loop header, bb2 the loop body.

In this case, when we run markBlock() over bb1, we first find bb2,
post-dominated by bb1. We then walk bb2 (pushing it to worklist), where we find
bb1 as its predecessor -- in that case we must mark the term inst of bb1 (a
cond_br), so that the loop count variable gets promoted to tensor.

The comment blocks in the code and added test cases provide more context.

mhong commented:
@swift-ci please test tensorflow

rxwei reviewed (approved):
Thanks for the fix!

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -763,6 +764,13 @@ public extension _TensorComputation {
+    } else {
+      internalConsistencyCheck(
+        _RuntimeConfig.usesSynchronousExecution,
+        """
```
Put this `"""` at the end of the last line.

mhong replied:
Done.

rxwei replied:
Did you move the closing delimiter instead of just moving the starting delimiter?

The closing delimiter should start on a new line, but the starting delimiter can be moved.

mhong replied:
Yes this works. Thanks!

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -763,6 +764,13 @@ public extension _TensorComputation {
       internalConsistencyCheck(joinStatus == 0)
       self.pthread = nil
+    } else {
+      internalConsistencyCheck(
```
Is the purpose of this mainly to prevent intermediate changes to `_RuntimeConfig.usesSynchronousExecution`?

Changing this global variable at any time between Start and End should be considered inconsistency or UB. In that case, we could add something to `willSet` to crash if it's set in between the start and end of any computation.

mhong replied:
The sanity check here focuses on `pthread` -- we want to make sure that when `finish()` is called, `pthread` is not nil. But we make an exception for the case of sync execution, where `pthread` is can legitimately be nil here.

lattner commented:
I'm not following all of the logic in your change, but you're right that there is a bug here.

The 'if (numTensorSuccs == 1) {' case was supposed to be an compile time optimization that shouldn't theoretically be necessary.  Does your testcase get fixed if you disable it?

I'm going to keep looking at your patch to try to understand what is going on, but from what I can tell, bb3 is not included in tensorCodeBlocks?

lattner commented:
This code *should* be adding BB3 to the set:

 // In addition to blocks with ops in them, we also add any block that has a
  // return, since that will be the head of the post dominator tree, and it is
  // possible that there are no ops in that block.
  for (auto &bb : fn.getBlocks())
    if (isa<ReturnInst>(bb.getTerminator()))
      worklist.push_back(&bb);

mhong commented:
@lattner Right bb3 is added.

After `tensorCodeBlocks.compute(tensorOps)`, we have all 4 BBs in the tensor code blocks.

```
(gdb) p tensorCodeBlocks.nodes
$60 = std::vector of length 4, capacity 4 = {{BB = 0x8a29528, Parent = 0x7fffffff9ad8, 
    Predecessors = std::vector of length 1, capacity 1 = {0x8f8ee40}, 
    Successors = std::vector of length 0, capacity 0}, {BB = 0x8dfe578, 
    Parent = 0x7fffffff9ad8, Predecessors = std::vector of length 2, capacity 2 = {
      0x8f8eec0, 0x8f8ee80}, Successors = std::vector of length 2, capacity 2 = {
      0x8f8ee80, 0x8f8ee00}}, {BB = 0x8a29570, Parent = 0x7fffffff9ad8, 
    Predecessors = std::vector of length 1, capacity 1 = {0x8f8ee40}, 
    Successors = std::vector of length 1, capacity 1 = {0x8f8ee40}}, {BB = 0x8dfdff0, 
    Parent = 0x7fffffff9ad8, Predecessors = std::vector of length 0, capacity 0, 
    Successors = std::vector of length 1, capacity 1 = {0x8f8ee40}}}
```

Also the `INPUT FUNCTION` is:

```
---- INPUT FUNCTION $S15sends_recvs_tmp6SR8191yyF ----------
// SR8191()
sil @$S15sends_recvs_tmp6SR8191yyF : $@convention(thin) () -> () {
bb0:
  %0 = float_literal $Builtin.FPIEEE32, 0x3F800000 // 1 // user: %1
  %1 = builtin "__tfop_tfc.scalarToTensor,$in"(%0 : $Builtin.FPIEEE32) : $TensorHandle<Float> // users: %63, %24, %23, %22, %22, %21, %20, %3, %2
  strong_retain %1 : $TensorHandle<Float>         // id: %2
  strong_release %1 : $TensorHandle<Float>        // id: %3
  %4 = integer_literal $Builtin.Int64, 0          // user: %18
  %5 = metatype $@thick Float.Type                // user: %22
  %6 = integer_literal $Builtin.Word, 1           // user: %31
  %7 = integer_literal $Builtin.Int64, 1          // users: %55, %8
  %8 = struct $Int (%7 : $Builtin.Int64)          // user: %16
  // function_ref default argument 1 of print(_:separator:terminator:)
  %9 = function_ref @$Ss5print_9separator10terminatoryypd_S2StFfA0_ : $@convention(thin) () -> @owned String // user: %43
  // function_ref default argument 2 of print(_:separator:terminator:)
  %10 = function_ref @$Ss5print_9separator10terminatoryypd_S2StFfA1_ : $@convention(thin) () -> @owned String // user: %44
  // function_ref print(_:separator:terminator:)
  %11 = function_ref @$Ss5print_9separator10terminatoryypd_S2StF : $@convention(thin) (@guaranteed Array<Any>, @guaranteed String, @guaranteed String) -> () // user: %45
  %12 = integer_literal $Builtin.Int1, -1         // user: %55
  %13 = integer_literal $Builtin.Int64, 10        // user: %60
  %14 = integer_literal $Builtin.Int64, 2         // user: %15
  %15 = struct $UInt (%14 : $Builtin.Int64)       // user: %16
  %16 = struct $_SwiftArrayBodyStorage (%8 : $Int, %15 : $UInt) // user: %17
  %17 = struct $_ArrayBody (%16 : $_SwiftArrayBodyStorage) // user: %34
  br bb1(%4 : $Builtin.Int64)                     // id: %18

// %19                                            // user: %55
bb1(%19 : $Builtin.Int64):                        // Preds: bb0 bb2
  strong_retain %1 : $TensorHandle<Float>         // id: %20
  strong_retain %1 : $TensorHandle<Float>         // id: %21
  %22 = builtin "__tfop_Add,$in,$in,T"(%1 : $TensorHandle<Float>, %1 : $TensorHandle<Float>, %5 : $@thick Float.Type) : $TensorHandle<Float> // users: %28, %59, %41, %29, %27, %25, %26
  strong_release %1 : $TensorHandle<Float>        // id: %23
  strong_release %1 : $TensorHandle<Float>        // id: %24
  strong_retain %22 : $TensorHandle<Float>        // id: %25
  %26 = struct $Tensor<Float> (%22 : $TensorHandle<Float>) // users: %30, %42
  strong_retain %22 : $TensorHandle<Float>        // id: %27
  strong_release %22 : $TensorHandle<Float>       // id: %28
  strong_release %22 : $TensorHandle<Float>       // id: %29
  debug_value %26 : $Tensor<Float>, let, name "y" // id: %30
  %31 = alloc_ref [tail_elems $Any * %6 : $Builtin.Word] $_ContiguousArrayStorage<Any> // user: %32
  %32 = upcast %31 : $_ContiguousArrayStorage<Any> to $_ContiguousArrayStorageBase // users: %38, %39, %54, %37, %35, %33
  %33 = ref_element_addr %32 : $_ContiguousArrayStorageBase, #_ContiguousArrayStorageBase.countAndCapacity // user: %34
  store %17 to %33 : $*_ArrayBody                 // id: %34
  %35 = struct $_ContiguousArrayBuffer<Any> (%32 : $_ContiguousArrayStorageBase) // user: %36
  %36 = struct $Array<Any> (%35 : $_ContiguousArrayBuffer<Any>) // user: %45
  %37 = ref_tail_addr %32 : $_ContiguousArrayStorageBase, $Any // user: %40
  strong_retain %32 : $_ContiguousArrayStorageBase // id: %38
  strong_release %32 : $_ContiguousArrayStorageBase // id: %39
  %40 = init_existential_addr %37 : $*Any, $Tensor<Float> // user: %42
  strong_retain %22 : $TensorHandle<Float>        // id: %41
  store %26 to %40 : $*Tensor<Float>              // id: %42
  %43 = apply %9() : $@convention(thin) () -> @owned String // users: %50, %45
  %44 = apply %10() : $@convention(thin) () -> @owned String // users: %46, %45
  %45 = apply %11(%36, %43, %44) : $@convention(thin) (@guaranteed Array<Any>, @guaranteed String, @guaranteed String) -> ()
  %46 = struct_extract %44 : $String, #String._guts // user: %47
  %47 = struct_extract %46 : $_StringGuts, #_StringGuts._object // user: %48
  %48 = struct_extract %47 : $_StringObject, #_StringObject._object // user: %49
  strong_release %48 : $Builtin.BridgeObject      // id: %49
  %50 = struct_extract %43 : $String, #String._guts // user: %51
  %51 = struct_extract %50 : $_StringGuts, #_StringGuts._object // user: %52
  %52 = struct_extract %51 : $_StringObject, #_StringObject._object // user: %53
  strong_release %52 : $Builtin.BridgeObject      // id: %53
  strong_release %32 : $_ContiguousArrayStorageBase // id: %54
  %55 = builtin "sadd_with_overflow_Int64"(%19 : $Builtin.Int64, %7 : $Builtin.Int64, %12 : $Builtin.Int1) : $(Builtin.Int64, Builtin.Int1) // users: %57, %56
  %56 = tuple_extract %55 : $(Builtin.Int64, Builtin.Int1), 0 // users: %60, %62
  %57 = tuple_extract %55 : $(Builtin.Int64, Builtin.Int1), 1 // user: %58
  cond_fail %57 : $Builtin.Int1                   // id: %58
  strong_release %22 : $TensorHandle<Float>       // id: %59
  %60 = builtin "cmp_slt_Int64"(%56 : $Builtin.Int64, %13 : $Builtin.Int64) : $Builtin.Int1 // user: %61
  cond_br %60, bb2, bb3                           // id: %61

bb2:                                              // Preds: bb1
  br bb1(%56 : $Builtin.Int64)                    // id: %62

bb3:                                              // Preds: bb1
  strong_release %1 : $TensorHandle<Float>        // id: %63
  %64 = tuple ()                                  // user: %65
  return %64 : $()                                // id: %65
} // end sil function '$S15sends_recvs_tmp6SR8191yyF'

---- END OF INPUT FUNCTION ----------
```

The main thing this patch is trying to do is that we mark the `cond_br` in bb1, and in turn that will mark the loop related blocks (bb2, bb3) for accelerator. Let me know if you need more info.

mhong commented:
All logic related to `numTensorSuccs` is sanity-check. The main behavior change in this patch is when `BB == pred` in `markBlock()` -- that happens when we:
1. enter `markBlock()` with `BB` set to `bb1` 
2. push bb2 onto `worklist`
3. pop it off into `thisBB`, and get bb1 again as `pred`.

In that case, previously we will skip processing `pred` because bb1 has been inserted into `visited` -- but we have not marked its termInst (`cond_br`)! In this patch, we check that if `BB == pred`, we will mark its predInst.

lattner commented:
I'm not able to repro the failure.  This generates a CFG that looks the same:

import TensorFlow

@inline(never)
func printTensor(_ t : Tensor<Float>) { print(t)}


public func foo() {
  let t = Tensor<Float>(1.0)
  var i = 0
  repeat {
    let y = t + t
    printTensor(y)
    i += 1
  } while i < 10
}

mhong commented:
with this test case:

```
public func SR8191() {
  let t = Tensor<Float>(1.0)
  var i = 0
  repeat {
    let y = t + t // expected-warning {{value implicitly copied to the host}}
    print(y)
    i += 1
  } while i < 10
}
```

On github tensorflow head:
```
hongm@hongm:~/ssd_part/git/swift-base/swift$ git br -vv
WARNING: terminal is not fully functional
-  (press RETURN)
  loop_bug     823d5b5565 Fixed a bug in TFFunctionPartition::markBlock() when handling a loop.
* tensorflow   c21f5b7e3e [apple/tensorflow] [strict-deabstraction] Fix a crash with array lists that have multiple (#17758)
```

I'm getting (note the markings and Accelerator function is incorrect):

```
----
---- INPUT FUNCTION $S15sends_recvs_tmp6SR8191yyF ----------
// SR8191()
sil @$S15sends_recvs_tmp6SR8191yyF : $@convention(thin) () -> () {
bb0:
  %0 = float_literal $Builtin.FPIEEE32, 0x3F800000 // 1 // user: %1
  %1 = builtin "__tfop_tfc.scalarToTensor,$in"(%0 : $Builtin.FPIEEE32) : $TensorHandle<Float> // users: %63, %24, %23, %22, %22, %21, %20, %3, %2
  strong_retain %1 : $TensorHandle<Float>         // id: %2
  strong_release %1 : $TensorHandle<Float>        // id: %3
  %4 = integer_literal $Builtin.Int64, 0          // user: %18
  %5 = metatype $@thick Float.Type                // user: %22
  %6 = integer_literal $Builtin.Word, 1           // user: %31
  %7 = integer_literal $Builtin.Int64, 1          // users: %55, %8
  %8 = struct $Int (%7 : $Builtin.Int64)          // user: %16
  // function_ref default argument 1 of print(_:separator:terminator:)
  %9 = function_ref @$Ss5print_9separator10terminatoryypd_S2StFfA0_ : $@convention(thin) () -> @owned String // user: %43
  // function_ref default argument 2 of print(_:separator:terminator:)
  %10 = function_ref @$Ss5print_9separator10terminatoryypd_S2StFfA1_ : $@convention(thin) () -> @owned String // user: %44
  // function_ref print(_:separator:terminator:)
  %11 = function_ref @$Ss5print_9separator10terminatoryypd_S2StF : $@convention(thin) (@guaranteed Array<Any>, @guaranteed String, @guaranteed String) -> () // user: %45
  %12 = integer_literal $Builtin.Int1, -1         // user: %55
  %13 = integer_literal $Builtin.Int64, 10        // user: %60
  %14 = integer_literal $Builtin.Int64, 2         // user: %15
  %15 = struct $UInt (%14 : $Builtin.Int64)       // user: %16
  %16 = struct $_SwiftArrayBodyStorage (%8 : $Int, %15 : $UInt) // user: %17
  %17 = struct $_ArrayBody (%16 : $_SwiftArrayBodyStorage) // user: %34
  br bb1(%4 : $Builtin.Int64)                     // id: %18

// %19                                            // user: %55
bb1(%19 : $Builtin.Int64):                        // Preds: bb0 bb2
  strong_retain %1 : $TensorHandle<Float>         // id: %20
  strong_retain %1 : $TensorHandle<Float>         // id: %21
  %22 = builtin "__tfop_Add,$in,$in,T"(%1 : $TensorHandle<Float>, %1 : $TensorHandle<Float>, %5 : $@thick Float.Type) : $TensorHandle<Float> // users: %28, %59, %41, %29, %27, %25, %26
  strong_release %1 : $TensorHandle<Float>        // id: %23
  strong_release %1 : $TensorHandle<Float>        // id: %24
  strong_retain %22 : $TensorHandle<Float>        // id: %25
  %26 = struct $Tensor<Float> (%22 : $TensorHandle<Float>) // users: %30, %42
  strong_retain %22 : $TensorHandle<Float>        // id: %27
  strong_release %22 : $TensorHandle<Float>       // id: %28
  strong_release %22 : $TensorHandle<Float>       // id: %29
  debug_value %26 : $Tensor<Float>, let, name "y" // id: %30
  %31 = alloc_ref [tail_elems $Any * %6 : $Builtin.Word] $_ContiguousArrayStorage<Any> // user: %32
  %32 = upcast %31 : $_ContiguousArrayStorage<Any> to $_ContiguousArrayStorageBase // users: %38, %39, %54, %37, %35, %33
  %33 = ref_element_addr %32 : $_ContiguousArrayStorageBase, #_ContiguousArrayStorageBase.countAndCapacity // user: %34
  store %17 to %33 : $*_ArrayBody                 // id: %34
  %35 = struct $_ContiguousArrayBuffer<Any> (%32 : $_ContiguousArrayStorageBase) // user: %36
  %36 = struct $Array<Any> (%35 : $_ContiguousArrayBuffer<Any>) // user: %45
  %37 = ref_tail_addr %32 : $_ContiguousArrayStorageBase, $Any // user: %40
  strong_retain %32 : $_ContiguousArrayStorageBase // id: %38
  strong_release %32 : $_ContiguousArrayStorageBase // id: %39
  %40 = init_existential_addr %37 : $*Any, $Tensor<Float> // user: %42
  strong_retain %22 : $TensorHandle<Float>        // id: %41
  store %26 to %40 : $*Tensor<Float>              // id: %42
  %43 = apply %9() : $@convention(thin) () -> @owned String // users: %50, %45
  %44 = apply %10() : $@convention(thin) () -> @owned String // users: %46, %45
  %45 = apply %11(%36, %43, %44) : $@convention(thin) (@guaranteed Array<Any>, @guaranteed String, @guaranteed String) -> ()
  %46 = struct_extract %44 : $String, #String._guts // user: %47
  %47 = struct_extract %46 : $_StringGuts, #_StringGuts._object // user: %48
  %48 = struct_extract %47 : $_StringObject, #_StringObject._object // user: %49
  strong_release %48 : $Builtin.BridgeObject      // id: %49
  %50 = struct_extract %43 : $String, #String._guts // user: %51
  %51 = struct_extract %50 : $_StringGuts, #_StringGuts._object // user: %52
  %52 = struct_extract %51 : $_StringObject, #_StringObject._object // user: %53
  strong_release %52 : $Builtin.BridgeObject      // id: %53
  strong_release %32 : $_ContiguousArrayStorageBase // id: %54
  %55 = builtin "sadd_with_overflow_Int64"(%19 : $Builtin.Int64, %7 : $Builtin.Int64, %12 : $Builtin.Int1) : $(Builtin.Int64, Builtin.Int1) // users: %57, %56
  %56 = tuple_extract %55 : $(Builtin.Int64, Builtin.Int1), 0 // users: %60, %62
  %57 = tuple_extract %55 : $(Builtin.Int64, Builtin.Int1), 1 // user: %58
  cond_fail %57 : $Builtin.Int1                   // id: %58
  strong_release %22 : $TensorHandle<Float>       // id: %59
  %60 = builtin "cmp_slt_Int64"(%56 : $Builtin.Int64, %13 : $Builtin.Int64) : $Builtin.Int1 // user: %61
  cond_br %60, bb2, bb3                           // id: %61

bb2:                                              // Preds: bb1
  br bb1(%56 : $Builtin.Int64)                    // id: %62

bb3:                                              // Preds: bb1
  strong_release %1 : $TensorHandle<Float>        // id: %63
  %64 = tuple ()                                  // user: %65
  return %64 : $()                                // id: %65
} // end sil function '$S15sends_recvs_tmp6SR8191yyF'

---- END OF INPUT FUNCTION ----------

---- ANALYSIS STATE FOR FUNCTION $S15sends_recvs_tmp6SR8191yyF ----------
Tensor start point:   %0 = float_literal $Builtin.FPIEEE32, 0x3F800000 // 1 // user: %1
Tensor end point:   strong_release %1 : $TensorHandle<Float>        // id: %24
SIL with markings:

bb0:
[Move]	  %0 = float_literal $Builtin.FPIEEE32, 0x3F800000 // 1 // user: %1
[Move]	  %1 = builtin "__tfop_tfc.scalarToTensor,$in"(%0 : $Builtin.FPIEEE32) : $TensorHandle<Float> // users: %23, %23, %2, %3, %20, %21, %24, %25, %64
	  strong_retain %1 : $TensorHandle<Float>         // id: %2
	  strong_release %1 : $TensorHandle<Float>        // id: %3
	  %4 = integer_literal $Builtin.Int64, 0          // user: %18
	  %5 = metatype $@thick Float.Type                // user: %23
	  %6 = integer_literal $Builtin.Word, 1           // user: %32
	  %7 = integer_literal $Builtin.Int64, 1          // users: %56, %8
	  %8 = struct $Int (%7 : $Builtin.Int64)          // user: %16
	  // function_ref default argument 1 of print(_:separator:terminator:)
  %9 = function_ref @$Ss5print_9separator10terminatoryypd_S2StFfA0_ : $@convention(thin) () -> @owned String // user: %44
	  // function_ref default argument 2 of print(_:separator:terminator:)
  %10 = function_ref @$Ss5print_9separator10terminatoryypd_S2StFfA1_ : $@convention(thin) () -> @owned String // user: %45
	  // function_ref print(_:separator:terminator:)
  %11 = function_ref @$Ss5print_9separator10terminatoryypd_S2StF : $@convention(thin) (@guaranteed Array<Any>, @guaranteed String, @guaranteed String) -> () // user: %46
	  %12 = integer_literal $Builtin.Int1, -1         // user: %56
	  %13 = integer_literal $Builtin.Int64, 10        // user: %61
	  %14 = integer_literal $Builtin.Int64, 2         // user: %15
	  %15 = struct $UInt (%14 : $Builtin.Int64)       // user: %16
	  %16 = struct $_SwiftArrayBodyStorage (%8 : $Int, %15 : $UInt) // user: %17
	  %17 = struct $_ArrayBody (%16 : $_SwiftArrayBodyStorage) // user: %35
	  br bb1(%4 : $Builtin.Int64)                     // id: %18

bb1:
	%19 = argument of bb1 : $Builtin.Int64            // user: %56
	  strong_retain %1 : $TensorHandle<Float>         // id: %20
	  strong_retain %1 : $TensorHandle<Float>         // id: %21
	  %22 = string_literal utf8 "/device:CPU:0"       // user: %23
[Move]	  %23 = builtin "__tfop_Add,$in,$in,T,__device"(%1 : $TensorHandle<Float>, %1 : $TensorHandle<Float>, %5 : $@thick Float.Type, %22 : $Builtin.RawPointer) : $TensorHandle<Float> // users: %27, %26, %28, %30, %42, %60, %29
	  strong_release %1 : $TensorHandle<Float>        // id: %24
	  strong_release %1 : $TensorHandle<Float>        // id: %25
	  strong_retain %23 : $TensorHandle<Float>        // id: %26
	  %27 = struct $Tensor<Float> (%23 : $TensorHandle<Float>) // users: %31, %43
	  strong_retain %23 : $TensorHandle<Float>        // id: %28
	  strong_release %23 : $TensorHandle<Float>       // id: %29
	  strong_release %23 : $TensorHandle<Float>       // id: %30
	  debug_value %27 : $Tensor<Float>, let, name "y" // id: %31
	  %32 = alloc_ref [tail_elems $Any * %6 : $Builtin.Word] $_ContiguousArrayStorage<Any> // user: %33
	  %33 = upcast %32 : $_ContiguousArrayStorage<Any> to $_ContiguousArrayStorageBase // users: %39, %40, %55, %38, %36, %34
	  %34 = ref_element_addr %33 : $_ContiguousArrayStorageBase, #_ContiguousArrayStorageBase.countAndCapacity // user: %35
	  store %17 to %34 : $*_ArrayBody                 // id: %35
	  %36 = struct $_ContiguousArrayBuffer<Any> (%33 : $_ContiguousArrayStorageBase) // user: %37
	  %37 = struct $Array<Any> (%36 : $_ContiguousArrayBuffer<Any>) // user: %46
	  %38 = ref_tail_addr %33 : $_ContiguousArrayStorageBase, $Any // user: %41
	  strong_retain %33 : $_ContiguousArrayStorageBase // id: %39
	  strong_release %33 : $_ContiguousArrayStorageBase // id: %40
	  %41 = init_existential_addr %38 : $*Any, $Tensor<Float> // user: %43
	  strong_retain %23 : $TensorHandle<Float>        // id: %42
	  store %27 to %41 : $*Tensor<Float>              // id: %43
	  %44 = apply %9() : $@convention(thin) () -> @owned String // users: %51, %46
	  %45 = apply %10() : $@convention(thin) () -> @owned String // users: %47, %46
	  %46 = apply %11(%37, %44, %45) : $@convention(thin) (@guaranteed Array<Any>, @guaranteed String, @guaranteed String) -> ()
	  %47 = struct_extract %45 : $String, #String._guts // user: %48
	  %48 = struct_extract %47 : $_StringGuts, #_StringGuts._object // user: %49
	  %49 = struct_extract %48 : $_StringObject, #_StringObject._object // user: %50
	  strong_release %49 : $Builtin.BridgeObject      // id: %50
	  %51 = struct_extract %44 : $String, #String._guts // user: %52
	  %52 = struct_extract %51 : $_StringGuts, #_StringGuts._object // user: %53
	  %53 = struct_extract %52 : $_StringObject, #_StringObject._object // user: %54
	  strong_release %53 : $Builtin.BridgeObject      // id: %54
	  strong_release %33 : $_ContiguousArrayStorageBase // id: %55
	  %56 = builtin "sadd_with_overflow_Int64"(%19 : $Builtin.Int64, %7 : $Builtin.Int64, %12 : $Builtin.Int1) : $(Builtin.Int64, Builtin.Int1) // users: %58, %57
	  %57 = tuple_extract %56 : $(Builtin.Int64, Builtin.Int1), 0 // users: %61, %63
	  %58 = tuple_extract %56 : $(Builtin.Int64, Builtin.Int1), 1 // user: %59
	  cond_fail %58 : $Builtin.Int1                   // id: %59
	  strong_release %23 : $TensorHandle<Float>       // id: %60
	  %61 = builtin "cmp_slt_Int64"(%57 : $Builtin.Int64, %13 : $Builtin.Int64) : $Builtin.Int1 // user: %62
	  cond_br %61, bb2, bb3                           // id: %62

bb2:
	  br bb1(%57 : $Builtin.Int64)                    // id: %63

bb3:
	  strong_release %1 : $TensorHandle<Float>        // id: %64
	  %65 = tuple ()                                  // user: %66
	  return %65 : $()                                // id: %66
---- END OF ANALYSIS STATE FOR FUNCTION ----------

---- PARTITION STATE FOR FUNCTION $S15sends_recvs_tmp6SR8191yyF ----------
(Possibly updated) tensor end point:   strong_release %1 : $TensorHandle<Float>        // id: %24
There are 1 result values:
  %23 = builtin "__tfop_Add,$in,$in,T,__device"(%1 : $TensorHandle<Float>, %1 : $TensorHandle<Float>, %5 : $@thick Float.Type, %22 : $Builtin.RawPointer) : $TensorHandle<Float> // users: %27, %26, %28, %30, %42, %60, %29
---- END OF PARTITION STATE FOR FUNCTION ----------

--- TFPartition Accelerator Result: $S15sends_recvs_tmp6SR8191yyF.tf
// SR8191()
sil private @$S15sends_recvs_tmp6SR8191yyF.tf : $@callee_owned () -> TensorHandle<Float> {
bb0:
  %0 = float_literal $Builtin.FPIEEE32, 0x3F800000 // 1 // user: %3
  %1 = integer_literal $Builtin.Int32, 1          // user: %3
  %2 = string_literal utf8 "ALL_DEVICES"          // user: %3
  %3 = builtin "__tfop_Const,dtype$dtype,value$tensor,__device"(%1 : $Builtin.Int32, %0 : $Builtin.FPIEEE32, %2 : $Builtin.RawPointer) : $TensorHandle<Builtin.FPIEEE32> // user: %4
  %4 = unchecked_ref_cast %3 : $TensorHandle<Builtin.FPIEEE32> to $TensorHandle<Float> // users: %7, %7
  %5 = metatype $@thick Float.Type                // user: %7
  %6 = string_literal utf8 "/device:CPU:0"        // user: %7
  %7 = builtin "__tfop_Add,$in,$in,T,__device"(%4 : $TensorHandle<Float>, %4 : $TensorHandle<Float>, %5 : $@thick Float.Type, %6 : $Builtin.RawPointer) : $TensorHandle<Float> // user: %8
  return %7 : $TensorHandle<Float>                // id: %8
} // end sil function '$S15sends_recvs_tmp6SR8191yyF.tf'

----
--- XLA CFG Canonicalize: $S15sends_recvs_tmp6SR8191yyF.tf
block bb0
--- XLA CFG Canonicalize end
----
--- TFDevicePartition Cross Device Tensor Transfer Annotation Result: $S15sends_recvs_tmp6SR8191yyF.tf
// SR8191()
sil private @$S15sends_recvs_tmp6SR8191yyF.tf : $@callee_owned () -> TensorHandle<Float> {
bb0:
  %0 = float_literal $Builtin.FPIEEE32, 0x3F800000 // 1 // user: %3
  %1 = integer_literal $Builtin.Int32, 1          // user: %3
  %2 = string_literal utf8 "ALL_DEVICES"          // user: %3
  %3 = builtin "__tfop_Const,dtype$dtype,value$tensor,__device"(%1 : $Builtin.Int32, %0 : $Builtin.FPIEEE32, %2 : $Builtin.RawPointer) : $TensorHandle<Builtin.FPIEEE32> // user: %4
  %4 = unchecked_ref_cast %3 : $TensorHandle<Builtin.FPIEEE32> to $TensorHandle<Float> // users: %7, %7
  %5 = metatype $@thick Float.Type                // user: %7
  %6 = string_literal utf8 "/device:CPU:0"        // user: %7
  %7 = builtin "__tfop_Add,$in,$in,T,__device"(%4 : $TensorHandle<Float>, %4 : $TensorHandle<Float>, %5 : $@thick Float.Type, %6 : $Builtin.RawPointer) : $TensorHandle<Float> // user: %8
  return %7 : $TensorHandle<Float>                // id: %8
} // end sil function '$S15sends_recvs_tmp6SR8191yyF.tf'

----
--- TFDevicePartition Per-Device Function Extraction Result: $S15sends_recvs_tmp6SR8191yyF.tf_CPU.device_partition
// SR8191()
sil private @$S15sends_recvs_tmp6SR8191yyF.tf_CPU.device_partition : $@callee_owned () -> TensorHandle<Float> {
bb0:
  %0 = integer_literal $Builtin.Int32, 1          // user: %3
  %1 = float_literal $Builtin.FPIEEE32, 0x3F800000 // 1 // user: %3
  %2 = string_literal utf8 "ALL_DEVICES"          // user: %3
  %3 = builtin "__tfop_Const,dtype$dtype,value$tensor,__device"(%0 : $Builtin.Int32, %1 : $Builtin.FPIEEE32, %2 : $Builtin.RawPointer) : $TensorHandle<Builtin.FPIEEE32> // user: %4
  %4 = unchecked_ref_cast %3 : $TensorHandle<Builtin.FPIEEE32> to $TensorHandle<Float> // users: %7, %7
  %5 = metatype $@thick Float.Type                // user: %7
  %6 = string_literal utf8 "/device:CPU:0"        // user: %7
  %7 = builtin "__tfop_Add,$in,$in,T,__device"(%4 : $TensorHandle<Float>, %4 : $TensorHandle<Float>, %5 : $@thick Float.Type, %6 : $Builtin.RawPointer) : $TensorHandle<Float> // user: %8
  return %7 : $TensorHandle<Float>                // id: %8
} // end sil function '$S15sends_recvs_tmp6SR8191yyF.tf_CPU.device_partition'

----
wrote binary graph of 557 bytes to '/tmp/tf-dump-graph-$S15sends_recvs_tmp6SR8191yyF.tf-b79577.pb'
wrote textual graph of 1352 bytes to '/tmp/tf-dump-graph-$S15sends_recvs_tmp6SR8191yyF.tf-1b31e3.pbtxt'
The graph proto is: 
node {
  name: "tfc_func_S15sends_recvs_tmp6SR8191yyF.tf"
  op: "S15sends_recvs_tmp6SR8191yyF.tf_CPU.device_partition"
  device: "/device:CPU:0"
}
node {
  name: "tfc_output_0_S15sends_recvs_tmp6SR8191yyF.tf"
  op: "Identity"
  input: "tfc_func_S15sends_recvs_tmp6SR8191yyF.tf"
  attr {
    key: "T"
    value {
      type: DT_FLOAT
    }
  }
}
library {
  function {
    signature {
      name: "S15sends_recvs_tmp6SR8191yyF.tf_CPU.device_partition"
      output_arg {
        name: "op_sr8191___1_158_13"
        type: DT_FLOAT
      }
    }
    node_def {
      name: "op/SR8191...158.13"
      op: "Const"
      device: "/device:CPU:0"
      attr {
        key: "dtype"
        value {
          type: DT_FLOAT
        }
      }
      attr {
        key: "value"
        value {
          tensor {
            dtype: DT_FLOAT
            tensor_shape {
            }
            float_val: 1
          }
        }
      }
    }
    node_def {
      name: "op/SR8191.._1.158.13"
      op: "Add"
      input: "op/SR8191...158.13:output:0"
      input: "op/SR8191...158.13:output:0"
      device: "/device:CPU:0"
      attr {
        key: "T"
        value {
          type: DT_FLOAT
        }
      }
    }
    ret {
      key: "op_sr8191___1_158_13"
      value: "op/SR8191.._1.158.13:z:0"
    }
  }
}
versions {
  producer: 26
  min_consumer: 12
}

--- TFPartition Host Result: $S15sends_recvs_tmp6SR8191yyF
// SR8191()
sil @$S15sends_recvs_tmp6SR8191yyF : $@convention(thin) () -> () {
bb0:
  %0 = string_literal bytes "0A6F0A287466635F66756E635F53313573656E64735F72656376735F746D70365352383139317979462E7466123453313573656E64735F72656376735F746D70365352383139317979462E74665F4350552E6465766963655F706172746974696F6E220D2F6465766963653A4350553A300A6B0A2C7466635F6F75747075745F305F53313573656E64735F72656376735F746D70365352383139317979462E746612084964656E746974791A287466635F66756E635F53313573656E64735F72656376735F746D70365352383139317979462E74662A070A01541202300112C6020AC3020A500A3453313573656E64735F72656376735F746D70365352383139317979462E74665F4350552E6465766963655F706172746974696F6E1A180A146F705F7372383139315F5F5F315F3135385F313318011A4E0A126F702F5352383139312E2E2E3135382E31331205436F6E7374220D2F6465766963653A4350553A302A150A0576616C7565120C420A080112002A040000803F2A0B0A056474797065120230011A6D0A146F702F5352383139312E2E5F312E3135382E313312034164641A1B6F702F5352383139312E2E2E3135382E31333A6F75747075743A301A1B6F702F5352383139312E2E2E3135382E31333A6F75747075743A30220D2F6465766963653A4350553A302A070A01541202300122300A146F705F7372383139315F5F5F315F3135385F313312186F702F5352383139312E2E5F312E3135382E31333A7A3A302204081A100C" // user: %4
  %1 = integer_literal $Builtin.Int64, 557        // user: %6
  %2 = string_literal utf8 "S15sends_recvs_tmp6SR8191yyF.tf" // user: %5
  %3 = integer_literal $Builtin.Int64, 0          // user: %7
  %4 = struct $UnsafeRawPointer (%0 : $Builtin.RawPointer) // user: %17
  %5 = struct $UnsafePointer<Int8> (%2 : $Builtin.RawPointer) // user: %17
  %6 = struct $Int (%1 : $Builtin.Int64)          // user: %17
  %7 = struct $Int (%3 : $Builtin.Int64)          // user: %17
  %8 = alloc_stack $OpaquePointer                 // users: %19, %9
  %9 = begin_access [read] [static] %8 : $*OpaquePointer // users: %18, %10
  %10 = address_to_pointer %9 : $*OpaquePointer to $Builtin.RawPointer // user: %11
  %11 = struct $UnsafePointer<OpaquePointer> (%10 : $Builtin.RawPointer) // user: %17
  %12 = integer_literal $Builtin.Int64, 0         // user: %13
  %13 = struct $Int (%12 : $Builtin.Int64)        // user: %17
  %14 = integer_literal $Builtin.Int64, 1         // user: %15
  %15 = struct $Int (%14 : $Builtin.Int64)        // users: %40, %17
  // function_ref _swift_tfc_StartTensorComputation
  %16 = function_ref @_swift_tfc_StartTensorComputation : $@convention(thin) (UnsafeRawPointer, Int, UnsafePointer<Int8>, UnsafePointer<OpaquePointer>, Int, Int, Int) -> @owned _TensorComputation // user: %17
  %17 = apply %16(%4, %6, %5, %11, %13, %7, %15) : $@convention(thin) (UnsafeRawPointer, Int, UnsafePointer<Int8>, UnsafePointer<OpaquePointer>, Int, Int, Int) -> @owned _TensorComputation // user: %40
  end_access %9 : $*OpaquePointer                 // id: %18
  dealloc_stack %8 : $*OpaquePointer              // id: %19
  %20 = integer_literal $Builtin.Int64, 0         // user: %33
  %21 = integer_literal $Builtin.Word, 1          // user: %53
  %22 = integer_literal $Builtin.Int64, 1         // users: %77, %23
  %23 = struct $Int (%22 : $Builtin.Int64)        // user: %31
  // function_ref default argument 1 of print(_:separator:terminator:)
  %24 = function_ref @$Ss5print_9separator10terminatoryypd_S2StFfA0_ : $@convention(thin) () -> @owned String // user: %65
  // function_ref default argument 2 of print(_:separator:terminator:)
  %25 = function_ref @$Ss5print_9separator10terminatoryypd_S2StFfA1_ : $@convention(thin) () -> @owned String // user: %66
  // function_ref print(_:separator:terminator:)
  %26 = function_ref @$Ss5print_9separator10terminatoryypd_S2StF : $@convention(thin) (@guaranteed Array<Any>, @guaranteed String, @guaranteed String) -> () // user: %67
  %27 = integer_literal $Builtin.Int1, -1         // user: %77
  %28 = integer_literal $Builtin.Int64, 10        // user: %82
  %29 = integer_literal $Builtin.Int64, 2         // user: %30
  %30 = struct $UInt (%29 : $Builtin.Int64)       // user: %31
  %31 = struct $_SwiftArrayBodyStorage (%23 : $Int, %30 : $UInt) // user: %32
  %32 = struct $_ArrayBody (%31 : $_SwiftArrayBodyStorage) // user: %56
  br bb1(%20 : $Builtin.Int64)                    // id: %33

// %34                                            // user: %77
bb1(%34 : $Builtin.Int64):                        // Preds: bb0 bb2
  %35 = alloc_stack $OpaquePointer                // users: %46, %42, %36
  %36 = begin_access [modify] [static] %35 : $*OpaquePointer // users: %41, %37
  %37 = address_to_pointer %36 : $*OpaquePointer to $Builtin.RawPointer // user: %38
  %38 = struct $UnsafeMutablePointer<OpaquePointer> (%37 : $Builtin.RawPointer) // user: %40
  // function_ref _swift_tfc_FinishTensorComputation
  %39 = function_ref @_swift_tfc_FinishTensorComputation : $@convention(thin) (@guaranteed _TensorComputation, UnsafeMutablePointer<OpaquePointer>, Int) -> () // user: %40
  %40 = apply %39(%17, %38, %15) : $@convention(thin) (@guaranteed _TensorComputation, UnsafeMutablePointer<OpaquePointer>, Int) -> ()
  end_access %36 : $*OpaquePointer                // id: %41
  %42 = load %35 : $*OpaquePointer                // user: %45
  %43 = alloc_ref $TensorHandle<Float>            // users: %50, %81, %63, %51, %49, %47, %48, %44
  %44 = ref_element_addr %43 : $TensorHandle<Float>, #TensorHandle.cTensorHandle // user: %45
  store %42 to %44 : $*OpaquePointer              // id: %45
  dealloc_stack %35 : $*OpaquePointer             // id: %46
  strong_retain %43 : $TensorHandle<Float>        // id: %47
  %48 = struct $Tensor<Float> (%43 : $TensorHandle<Float>) // users: %52, %64
  strong_retain %43 : $TensorHandle<Float>        // id: %49
  strong_release %43 : $TensorHandle<Float>       // id: %50
  strong_release %43 : $TensorHandle<Float>       // id: %51
  debug_value %48 : $Tensor<Float>, let, name "y" // id: %52
  %53 = alloc_ref [tail_elems $Any * %21 : $Builtin.Word] $_ContiguousArrayStorage<Any> // user: %54
  %54 = upcast %53 : $_ContiguousArrayStorage<Any> to $_ContiguousArrayStorageBase // users: %60, %61, %76, %59, %57, %55
  %55 = ref_element_addr %54 : $_ContiguousArrayStorageBase, #_ContiguousArrayStorageBase.countAndCapacity // user: %56
  store %32 to %55 : $*_ArrayBody                 // id: %56
  %57 = struct $_ContiguousArrayBuffer<Any> (%54 : $_ContiguousArrayStorageBase) // user: %58
  %58 = struct $Array<Any> (%57 : $_ContiguousArrayBuffer<Any>) // user: %67
  %59 = ref_tail_addr %54 : $_ContiguousArrayStorageBase, $Any // user: %62
  strong_retain %54 : $_ContiguousArrayStorageBase // id: %60
  strong_release %54 : $_ContiguousArrayStorageBase // id: %61
  %62 = init_existential_addr %59 : $*Any, $Tensor<Float> // user: %64
  strong_retain %43 : $TensorHandle<Float>        // id: %63
  store %48 to %62 : $*Tensor<Float>              // id: %64
  %65 = apply %24() : $@convention(thin) () -> @owned String // users: %72, %67
  %66 = apply %25() : $@convention(thin) () -> @owned String // users: %68, %67
  %67 = apply %26(%58, %65, %66) : $@convention(thin) (@guaranteed Array<Any>, @guaranteed String, @guaranteed String) -> ()
  %68 = struct_extract %66 : $String, #String._guts // user: %69
  %69 = struct_extract %68 : $_StringGuts, #_StringGuts._object // user: %70
  %70 = struct_extract %69 : $_StringObject, #_StringObject._object // user: %71
  strong_release %70 : $Builtin.BridgeObject      // id: %71
  %72 = struct_extract %65 : $String, #String._guts // user: %73
  %73 = struct_extract %72 : $_StringGuts, #_StringGuts._object // user: %74
  %74 = struct_extract %73 : $_StringObject, #_StringObject._object // user: %75
  strong_release %74 : $Builtin.BridgeObject      // id: %75
  strong_release %54 : $_ContiguousArrayStorageBase // id: %76
  %77 = builtin "sadd_with_overflow_Int64"(%34 : $Builtin.Int64, %22 : $Builtin.Int64, %27 : $Builtin.Int1) : $(Builtin.Int64, Builtin.Int1) // users: %79, %78
  %78 = tuple_extract %77 : $(Builtin.Int64, Builtin.Int1), 0 // users: %82, %84
  %79 = tuple_extract %77 : $(Builtin.Int64, Builtin.Int1), 1 // user: %80
  cond_fail %79 : $Builtin.Int1                   // id: %80
  strong_release %43 : $TensorHandle<Float>       // id: %81
  %82 = builtin "cmp_slt_Int64"(%78 : $Builtin.Int64, %28 : $Builtin.Int64) : $Builtin.Int1 // user: %83
  cond_br %82, bb2, bb3                           // id: %83

bb2:                                              // Preds: bb1
  br bb1(%78 : $Builtin.Int64)                    // id: %84

bb3:                                              // Preds: bb1
  %85 = tuple ()                                  // user: %86
  return %85 : $()                                // id: %86
} // end sil function '$S15sends_recvs_tmp6SR8191yyF'
```

mhong commented:
When you compile and this test case, it'll crash, because `@_swift_tfc_FinishTensorComputation()` is inside the loop body on the host function

lattner commented:
ah, I can repro the bug.

lattner commented:
Yes, you're right, we're not marking the condbr.

Commit:
Incorporated suggested changes from @lattner.

Also renamed `visited` to `addedToList` to better reflect its semantics, and
moved around sanity check code on `numTensorSuccs` to make it clear that the
core marking logic does not depend on it.

mhong commented:
@swift-ci please test tensorflow

mhong commented:
Incorporated suggested changes from @lattner. PTAL.

I'll address comments from @rxwei once we converge on the core marking algorithm.

mhong commented:
Test passed again: https://ci-external.swift.org/view/Pull%20Request/job/swift-PR-TensorFlow-Linux/295/

lattner commented on lib/SILOptimizer/Mandatory/TFPartition.cpp:
```diff
@@ -997,27 +1003,28 @@ void TFFunctionPartition::markBlock(SILBasicBlock *BB) {
+      // In those cases, continue walking `pred` if we haven't processed it,
+      // since it may be control-dependent on something.
+      if (tensorCodeBlocks.properlyPostDominates(BB, pred)) {
+        if (!addedToList.count(pred)) {
```
doing count followed by insert does two hash table lookups.  It is better to just do "insert" and check the bool result to see if it was already there.

mhong replied:
Good point. changed.

lattner commented:
I haven't had time to follow this today.  I'll try to take a look in more detail when I can, but otherwise we can talk on monday.

Commit:
Addressed comments from @lattner and @rxwei.

Commit:
Moved the start """ to a previous line as suggested by @rxwei.

mhong commented:
Closed by accident.

lattner reviewed (approved):
Awesome, thank you for explaining this to me!

Commit:
Refreshed the PR with the head of tensorflow branch, and updated comment to
reflect discussion with @lattner.

mhong commented:
@swift-ci please test tensorflow

mhong commented:
Thanks for the discussion! A quick summary:

- `markBlock()` does block level analysis, as an efficient implementation to *approximate* instruction level analysis. What we really care about is at the instruction level -- which ones should get marked for accelerator. In this case, when we have marked a tensor op, if it is control dependent on another inst (the term inst of some block), we must mark that inst a well.
- The block level analysis provides accuracy results to reflect control-dependence on the instruction level, (i.e., given two instructions `I1` and `I2`, `I1` is control dependent on `I2`, iff `block(I1)` is control dependent on `block(I2)`), except when the two instructions we are analyzing are in the same block. e.g.
```
bb1(...):
  %x = <some tensor inst>
  ...
   cond_br bb2, bb3
```
Here `%x` is control-dependent on `cond_br`, so we should mark the latter for accelerator. But since they are in the same block, the block is **not** control dependent on itself.

So the correct block-level analysis here is **proper** post-dominance -- when some block `B1` is a transitive predecessor of `B2`, and they are the same block, we know all tensor ops in that block are control-dependent on the term inst of that block.

mhong commented:
CI passed: https://ci-external.swift.org/view/Pull%20Request/job/swift-PR-TensorFlow-Linux/323/
